src/persist-client/src/internal/compact.rs

```diff
@@ -481,131 +505,131 @@ where
 
-            if updates == 0 {
-                continue;
-            }
-            // merge together parts and runs from each compaction round.
-            // parts are appended onto our existing vec, and then we shift
-            // the latest run offsets to account for prior parts.
-            //
-            // e.g. if we currently have 3 parts and 2 runs (including the implicit one from 0):
-            //         parts: [k0, k1, k2]
-            //         runs:  [    1     ]
-            //
-            // and we merge in another result with 2 parts and 2 runs:
-            //         parts: [k3, k4]
-            //         runs:  [    1]
-            //
-            // we our result will contain 5 parts and 4 runs:
-            //         parts: [k0, k1, k2, k3, k4]
-            //         runs:  [    1       3   4 ]
-            let run_offset = all_parts.len();
-            if all_parts.len() > 0 {
-                all_run_splits.push(run_offset);
+            // Split the runs into manageable chunks
+            let chunked_runs =
+                Self::chunk_runs(&ordered_runs, &cfg, &*metrics, run_reserved_memory_bytes);
+
+            let total_chunked_runs = chunked_runs.len();
+            let mut applied = 0;
+            for (runs, run_chunk_max_memory_usage) in chunked_runs {
+                metrics.compaction.chunks_compacted.inc();
+                metrics
+                    .compaction
+                    .runs_compacted
+                    .inc_by(u64::cast_from(runs.len()));
+
+                // Adjust parallelism based on how much memory we have left
+                let extra_outstanding_parts = (run_reserved_memory_bytes
+                    .saturating_sub(run_chunk_max_memory_usage))
+                    / cfg.batch.blob_target_size;
+
+                let mut run_cfg = cfg.clone();
+                run_cfg.batch.batch_builder_max_outstanding_parts = 1 + extra_outstanding_parts;
+
+                let batch = Self::compact_runs(
+                    &run_cfg,
+                    &req.shard_id,
+                    &req.desc,
+                    runs,
+                    Arc::clone(&blob),
+                    Arc::clone(&metrics),
+                    Arc::clone(&shard_metrics),
+                    Arc::clone(&isolated_runtime),
+                    write_schemas.clone(),
+                )
+                .await?;
+
+                let (parts, run_splits, run_meta, updates) =
+                    (batch.parts, batch.run_splits, batch.run_meta, batch.len);
+
+                assert!(
+                    (updates == 0 && parts.is_empty()) || (updates > 0 && !parts.is_empty()),
+                    "updates={}, parts={}",
+                    updates,
+                    parts.len(),
+                );
+
+                // Skip empty batches unless it's the last one
+                if updates == 0 && applied != total_chunked_runs - 1 {
```

Comment:
Also... are we guaranteed that `total_chunked_runs` is nonzero, or can this underflow?